pkg/slayers/path/scion/raw_spec.gobra

```diff
@@ -850,38 +844,52 @@ func XoverRightSeg(raw []byte, currInfIdx int, currHfIdx int, segs io.SegLens) {
 }
 
 ghost
-opaque
+requires  R55 < p
 requires  0 <= offset
-requires  0 <= currHFIdx && currHFIdx <= segLen
-requires  len(hops) == segLen - currHFIdx
+requires  0 <= currHfIdx && currHfIdx <= end
+requires  end <= segLen
 requires  offset + path.HopLen * segLen <= len(raw)
-requires  acc(sl.Bytes(raw, 0, len(raw)), R56)
-decreases
-pure func hopFieldsBytePositions(raw []byte, offset int, currHFIdx int, segLen int, hops seq[io.IO_HF]) bool {
-	return forall i int :: { hops[i] } 0 <= i && i < len(hops) ==>
-		hops[i] == path.BytesToIO_HF(raw, 0, offset + path.HopLen * (currHFIdx + i), len(raw))
+preserves acc(sl.Bytes(raw, 0, len(raw)), p)
+ensures   hopFields(raw, offset, currHfIdx, segLen)[:end-currHfIdx] ==
+	hopFields(raw, offset, currHfIdx, end)
+decreases end - currHfIdx
+func hopFieldsSliceEnd(raw []byte, offset int, currHfIdx int, segLen int, end int, p perm) {
+	if (currHfIdx != end) {
+		newP := (p + R55)/2
+		hopFieldsSliceEnd(raw, offset, currHfIdx + 1, segLen, end, newP)
+	}
 }
 
 ghost
 requires  R55 < p
 requires  0 <= offset
-requires  0 <= currHFIdx && currHFIdx <= segLen
+requires  0 <= start
+requires  0 <= currHfIdx && currHfIdx <= segLen - start
 requires  offset + path.HopLen * segLen <= len(raw)
 preserves acc(sl.Bytes(raw, 0, len(raw)), p)
-ensures   hopFieldsBytePositions(raw, offset, currHFIdx, segLen, hopFields(raw, offset, currHFIdx, segLen))
-decreases segLen - currHFIdx
-func hopFieldsBytePositionsLemma(
-	raw []byte,
-	offset int,
-	currHFIdx int,
-	segLen int,
-	p perm) {
-	newP := (p + R55)/2
-	hopfields := hopFields(raw, offset, currHFIdx, segLen)
-	if (currHFIdx != segLen) {
-		hopFieldsBytePositionsLemma(raw, offset, currHFIdx + 1, segLen, newP)
-		hopfieldsInc := hopFields(raw, offset, currHFIdx + 1, segLen)
-		assert reveal hopFieldsBytePositions(raw, offset, currHFIdx + 1, segLen, hopfieldsInc)
+ensures   hopFields(raw, offset, currHfIdx, segLen)[start:] ==
+ 	hopFields(raw, offset, currHfIdx+start, segLen)
+decreases start
+func hopFieldsSliceStart(raw []byte, offset int, currHfIdx int, segLen int, start int, p perm) {
```

Comment:
let's rename this lemma and introduce a public lemma that calls without taking a `perm` param